File: app/src/main/java/com/example/retube/fragment/HomeFragment.java
package com.example.retube.fragment;


import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.retube.R;
import com.example.retube.Retrofit.GetDataService;
import com.example.retube.Retrofit.RetrofitInstance;
import com.example.retube.activity.PlayActivity;
import com.example.retube.adapter.HomeAdapter;
import com.example.retube.models.Channel;
import com.example.retube.models.HomeMostPopular;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class HomeFragment extends Fragment {

    private HomeAdapter adapter;
    private LinearLayoutManager manager;
    private List<HomeMostPopular.Item> videoMostPopularList = new ArrayList<>();
    private HashMap<Integer, Channel.Item> channelList = new HashMap<Integer, Channel.Item>();
    public HomeFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(final LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home, container, false);
        // Inflate the layout for this fragment
        RecyclerView rv = view.findViewById(R.id.recyclerView);
        adapter = new HomeAdapter(getContext(),videoMostPopularList,channelList);
        manager = new LinearLayoutManager(getContext());
        rv.setAdapter(adapter);
        rv.setLayoutManager(manager);


        if(videoMostPopularList.size() == 0) {
            System.out.println("추가해야함");
            getVideoDetail();
        }

        adapter.setOnItemClickListener(new HomeAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(View v, int pos) {

                Intent intent = new Intent(getActivity(), PlayActivity.class);
                intent.putExtra("videoID",videoMostPopularList.get(pos).getId());
                startActivity(intent);

            }
        });



        return view;
    }


    private void getVideoDetail() {

        GetDataService dataService = RetrofitInstance.getRetrofit().create((GetDataService.class));
        final Call<HomeMostPopular> videoDetailRequest = dataService
                .getMostPopular("snippet,statistics","items(id,snippet(title,thumbnails,publishedAt,channelId,channelTitle),statistics)",
                        "mostPopular", getString(R.string.api_key),"KR",25);
        videoDetailRequest.enqueue(new Callback<HomeMostPopular>() {
            @Override
            public void onResponse(Call<HomeMostPopular> call, Response<HomeMostPopular> response) {

                if(response.isSuccessful()){
                    if(response.body()!=null){

                        videoMostPopularList.addAll(response.body().getItems());
                        System.out.println("성공"+videoMostPopularList.size());
                        getChannelThumbs();
                    }else{
                        System.out.println("실패");
                    }
                }else{
                    System.out.println("실패dd");
                }

            }

            @Override
            public void onFailure(Call<HomeMostPopular> call, Throwable t) {

            }
        });
    }

    private void getChannelThumbs(){
        for(int i=0;i<videoMostPopularList.size();i++){
            getChannelThumb(videoMostPopularList.get(i).getSnippet().getChannelId(),i);
        }
        System.out.println("이미지 불러오기 끝");

    }


    private void getChannelThumb(String id, final int pos){
        GetDataService dataService = RetrofitInstance.getRetrofit().create((GetDataService.class));
        final Call<Channel> channelListRequest = dataService
                .getChannels("snippet", id,
                        getString(R.string.api_key),10);
        channelListRequest.enqueue(new Callback<Channel>() {
            @Override
            public void onResponse(Call<Channel> call, Response<Channel> response) {

                if(response.isSuccessful()){
                    if(response.body()!=null){

                        System.out.println("이미지 불러오기 성공");
                        channelList.put(pos, response.body().getItems().get(0));

                        adapter.notifyDataSetChanged();
                        adapter.getItemCount();


                    }else{
                        System.out.println("실패");
                    }
                }else{
                    System.out.println("실패dd");
                }

            }

            @Override
            public void onFailure(Call<Channel> call, Throwable t) {

            }
        });
    }

}

File: app/src/main/java/com/example/retube/ui/home/HomeFragment.kt
package com.example.retube.ui.home

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.retube.R
import com.example.retube.activity.PlayActivity
import com.example.retube.data.network.YoutubeApi
import com.example.retube.data.repositories.YoutubeRepository
import com.example.retube.models.Channel
import com.example.retube.models.HomeMostPopular
import kotlinx.android.synthetic.main.fragment_home.*
import kotlinx.android.synthetic.main.row_item_home.*


class HomeFragment : Fragment() , RecyclerViewClickListener{

    private lateinit var factory: HomeViewModelFactory
    private lateinit var viewModel: HomeViewModel

    private val videoMostPopularList: MutableList<HomeMostPopular.Item> = mutableListOf()
    private var channelList: HashMap<Int, Channel.Item> = HashMap<Int, Channel.Item>()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_home, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        //리사이클러뷰 데이터 불러오기 -> 사진 썸네일 불러오기
        //클릭 리스너

        rc_home.layoutManager = LinearLayoutManager(requireContext())
        rc_home.setHasFixedSize(true)
        rc_home.adapter = HomeAdapter(videoMostPopularList,channelList,this)

        val api = YoutubeApi()
        val repository =
            YoutubeRepository(api)
        factory = HomeViewModelFactory(repository)
        viewModel = ViewModelProvider(this, factory).get(HomeViewModel::class.java)
        viewModel.getHomeDatas("snippet,statistics","items(id,snippet(title,thumbnails,publishedAt,channelId,channelTitle),statistics)",
            "mostPopular", getString(R.string.api_key),"KR",3)
        viewModel.homedatas.observe(viewLifecycleOwner, Observer { home ->

            videoMostPopularList.clear()
            videoMostPopularList.addAll(home.items)
            rc_home.adapter!!.notifyDataSetChanged()
            getAllChannelData()

        })

        viewModel.chennels.observe(viewLifecycleOwner, Observer { channel ->
            channelList.clear()
            channelList.putAll(channel)
            rc_home.adapter!!.notifyDataSetChanged()
        })


    }

    private fun getAllChannelData(){
        for (i in 0..videoMostPopularList.size-1){
            viewModel.getChannelDatas("snippet", videoMostPopularList.get(i).getSnippet().getChannelId(), getString(R.string.api_key),10 , i)
        }
    }

    override fun onRecyclerViewItemClick(view: View, pos: Int) {

        when(view){
            home_view -> {
                val intent = Intent(activity, PlayActivity::class.java)
                intent.putExtra("videoID", videoMostPopularList[pos].id)
                startActivity(intent)
            }
        }

    }

}